src/node_sqlite.h

```diff
@@ -317,6 +318,26 @@ class SQLTagStore : public BaseObject {
   friend class StatementExecutionHelper;
 };
 
+class AuthorizerFunction {
+ public:
+  AuthorizerFunction(Environment* env,
+                     v8::Local<v8::Function> fn,
+                     DatabaseSync* db);
+  ~AuthorizerFunction();
+  static int xAuthorizer(void* user_data,
+                         int action_code,
+                         const char* param1,
+                         const char* param2,
+                         const char* param3,
+                         const char* param4);
+  static void xDestroy(void* self);
+
+ private:
+  Environment* env_;
+  v8::Global<v8::Function> fn_;
```

Comment:
Using v8::Global is often a memory leak pitfall, and that seems to be the case here too. If this function holds a reference to the `DatabaseSync` instance (which it often will do, at least indirectly), no object here will ever end up being garbage collected.

You're probably better of turning this into a `BaseObject`, attaching the associated JS object to the `DatabaseSync` instance (maybe even in JS land) and then storing the function in an internal field rather than a `v8::Global<>` handle.

Reply:
Thanks for pointing this out! but I’m a bit confused. Why do you suggest turning this into a `BaseObject`? Wouldn’t it be sufficient to simply release the `AuthorizerFunction` handle when DatabaseSync is destroyed?

Reply:
@araujogui Right, but the `DatabaseSync` C++ object is destroyed when its associated JS object is garbage collected, and if the JS function that's attached to `AuthorizerFunction` has a reference to the `DatabaseSync` JS object (doesn't even have to be a direct one! It can just be that the `DatabaseSync` JS object is in one of the parents scope of the function), then none of these objects will be destroyed.

So you have

`AuthorizerFunction` → + `v8::Global<>` → JS function → DatabaseSync JS object → DatabaseSync C++ object → `AuthorizerFunction`

as a cyclic reference where the `v8::Global<>` is a GC root, and that means that none of these objects can be garbage collected

Reply:
Also, I was chatting with @geeksilva97 about this, we already have a similar problem with `UserDefinedFunction`.

Here, in this case, there might be an easier way than to add a `BaseObject` – we could just store the `AuthorizerFunction` in an internal slot of `DatabaseSync`, which removes the need for a separate class (and also the need to worry about deallocation at all).

That doesn't work for `UserDefinedFunction`, or at least not as easily, because we can have multiple `UserDefinedFunction` per `DatabaseSync` instance.

Reply:
@addaleax just out of curiosity, do you have code to reproduce that in the `UserDefinedFunction` case? I couldn't reproduce it, but I also don't know where the cycle comes from (the `DatabaseSync` C++ object doesn't hold a reference to the `UserDefinedFunction`). Here is the code I used:

```js
'use strict';
const { DatabaseSync } = require('node:sqlite');
let db = new DatabaseSync(':memory:');
db.function('custom', () => {
  console.log(db.isOpen); // Reference the db.
  return 42;
});
const r = db.prepare('SELECT custom() AS custom').get();
db.close();
db = null;

setInterval(() => {
  global.gc();
  console.log(r?.custom);
}, 1_000);
```

(I also added logs to the C++ code in the destructor for `UserDefinedFunction` and `DatabaseSync`, and both seemed to be called.)